woocommerce/woocommerce-ios
# Product Settings: show, edit and update Product Visibility

Resolves #2120


## Description
As part of Product Settings (part of Products Milestone 2), we want to allow a user to edit the product's visibility.
Turn out, that this task it’s a little bit trickier if compared to other Product Settings because we need to fetch the password from the WP.com Site Post API, implemented previously here: [Product Visibility Setting: Yosemite & Networking layers](https://github.com/woocommerce/woocommerce-ios/pull/2213), and the `password` is not part of Products, therefore we must update it consequently to the products, manage the changes of state, manage the state when the field has not been changed, pass the updated data back and forth along with the entire flow. The introduction of product visibility also changes the state's behavior: the state now contains only three types of states but can take the private state if changed by Product Visibility. In this case, the status field is disabled.

For @Garance91540: I used an icon for the password field that is different from the one initially chosen because the one on Figma was not available (not present on Gridicons, and I was unable to download it). I will update the PR with the correct icon as soon as I have it.

## Changes
There are a lot of changes, I'm sorry about this. Probably I could have split the logic and UI into two separate PRs.

- I updated `SitePostRemote` and `SitePostRemoteTests` because the path was wrong. In addition, `SitePostStore` has been added to` AuthenticatedStates`.
- New `passwordFieldImage` in `UIImage+Woo`
- New color `textBrand`, used in the password field.
- Removed `privateStatus` from the status selectable in the Product Status screen.
- `ProductSettingsRows`: managed the “status” cell (selectable or not), and added the new row Visibility
- `ProductSettingsViewController`: implemented the `PasswordCompletion` block, which is useful to pass back the original password when it’s fetched.
- `ProductSettingsViewModel`: fetch the product password, and implements a new callback `onPasswordRetrieved` on which `ProductSettingsViewController` will subscribe. Manage the unsaved changes.
- Implemented a new model `ProductVisibility` with all the options offered in the Product Visibility screen. This is an internal model, and we don’t need it in Networking or in Yosemite.
- `ProductVisibilityViewController`: it’s the view controller witch takes care of present the Product Visibility screen. 
- `ProductFormViewController`: now manage the password field. It uses `originalPassword` which will be the password fetched in Product Settings and a variable `password` which represents the edited password. Furthermore, the remote updating of the password takes place here, in cascade together with the product update, and only if the password has been populated (therefore if the user has opened the Product Settings). Manage also the unsaved changes when the password was edited.
- `SettingTitleAndValueTableViewCell` now implements the default style and another style which is the non-selectable style used by the product status row.
- `TitleAndTextFieldWithImageTableViewCell`: it’s a new cell that contains a title label, a text field, and an image. It’s used for the password field.
- `ProductSettings`: introduced the `password` variable.

## Testing

1) Navigate to a simple product
2) Navigate to Product Settings
3) Tap Visibility row
4) Select "Password Protected" -> The password row will appear, without the right icon.
5) Try to digit a password -> After the first letter, the right icon will appear.
6) Tap back -> The Visibility row now will show "Password Protected"
7) Try to save the product -> Verify that the password is showed also on the web product.
8) Close the product, reopen it. Open Product Settings -> Make sure the password is fetched correctly. You will see "Password Protected" in Visibility row.
9) Tap Visibility, and change it to private. Press back. -> Make sure "Status" row now shows "Private published" and it's not selectable.
10) Save the product -> Check on the web that the status is private.
11) Re-open Product Visibility, and Select Password Protected. Tap back without typing a password. -> The discard dialog (if you are not on iOS 13.4 https://github.com/woocommerce/woocommerce-ios/issues/2089) should appear.



## Screenshots

### Light Mode
| Public Visibility | Empty password | Password added | Discard changes | Visibility private
:-------------------------:|:-------------------------:|:-------------------------:|:-------------------------:|:-------------------------:
![Public visibility](https://user-images.githubusercontent.com/495617/81441488-274edb00-9172-11ea-8547-83851e9f5090.png) | ![Enter password empty](https://user-images.githubusercontent.com/495617/81441530-36ce2400-9172-11ea-9ef9-d3ff2cf85e8d.png) | ![Password added](https://user-images.githubusercontent.com/495617/81441568-49485d80-9172-11ea-852a-f8dc52141048.png) | ![Discard Visibility changes](https://user-images.githubusercontent.com/495617/81441619-5f561e00-9172-11ea-88d2-7b58ff669755.png) | ![Visibility private, status private](https://user-images.githubusercontent.com/495617/81441685-8280cd80-9172-11ea-96a4-378eb9faac17.png)

### Dark Mode
(I can't attach the dark mode version of the Discard Changes screen, because I need to run the app on iOS 13 simulator, which presents this bug https://github.com/woocommerce/woocommerce-ios/issues/2089).

| Public Visibility | Empty password | Password added | Visibility private
:-------------------------:|:-------------------------:|:-------------------------:|:-------------------------:
![public visibility 2](https://user-images.githubusercontent.com/495617/81442108-4601a180-9173-11ea-8e53-05ce0ec84301.png) | ![enter password 2](https://user-images.githubusercontent.com/495617/81442116-4863fb80-9173-11ea-9f1b-0a8cb7cfe210.png) | ![Password added 3](https://user-images.githubusercontent.com/495617/81442120-4a2dbf00-9173-11ea-82b0-1c0b471192b9.png) | ![Visibility private, status private 2](https://user-images.githubusercontent.com/495617/81442124-4bf78280-9173-11ea-8789-6c70cc822f0f.png)



Update release notes:
